src/mesh/NodeDB.cpp

```diff
@@ -718,6 +726,258 @@ void NodeDB::saveDeviceStateToDisk()
               &devicestate);
 }
 
+void NodeDB::initSavedMessages()
+{
+    FSCom.mkdir("/msgs");
+    for (int category = 0; category < CATEGORY_COUNT; category++)
+        FSCom.mkdir(("/msgs/" + std::to_string(category)).c_str());
+
+    updateMessageFileList();
+
+    updateMessageBounds();
+}
+
+bool removeDirectoryRecursively(const char *dirPath)
+{
+    File directory = FSCom.open(dirPath, FILE_O_WRITE);
+    if (!directory || !directory.isDirectory()) {
+        LOG_ERROR("Failed to open directory for removal\n");
+        return false;
+    }
+
+    File file = directory.openNextFile(FILE_O_WRITE);
+    while (file) {
+        String filePath = String(dirPath) + "/" + file.name();
+        if (file.isDirectory()) {
+            file.close();
+            if (!removeDirectoryRecursively(filePath.c_str())) {
+                LOG_ERROR("Failed to remove directory: %s\n", filePath);
+                return false;
+            }
+        } else {
+            file.close();
+            if (!FSCom.remove(filePath.c_str())) {
+                LOG_ERROR("Failed to remove file: %s\n", filePath);
+                return false;
+            }
+        }
+        file = directory.openNextFile(FILE_O_WRITE);
+    }
+
+    // Now that the directory is empty, it can be removed.
+    if (!FSCom.rmdir(dirPath)) {
+        LOG_ERROR("Failed to remove directory: %s\n", dirPath);
+        return false;
+    }
+
+    return true;
+}
+
+void NodeDB::clearSavedMessages()
+{
+    LOG_DEBUG("Clearing saved messages\n");
+
+    removeDirectoryRecursively("/msgs");
+
+    initSavedMessages();
+}
+
+bool NodeDB::messageIsDirectMessage(const meshtastic_MeshPacket &mp)
+{
+    if (mp.to == nodeDB->getNodeNum())
+        return true;
+
+    for (size_t i = 0; i < numMeshNodes; i++)
+        if (mp.to == nodeDB->getMeshNodeByIndex(i)->num)
+            return true;
+
+    return false;
+}
+
+void NodeDB::saveMessageToDisk(const meshtastic_MeshPacket &mp)
+{
+    auto *msg = new meshtastic_Message();
+    auto senderNode = nodeDB->getMeshNode(mp.from);
```

Comment:
Better to use `getMeshNode(getFrom(mp))` here, same for line 805. Currently if you're sending a packet from the phone to the device, `from` will still be 0 here, so it shows as "UNK".

Also currently packets sent from the phone will not be shown for DMs because `handleReceived()` is not called for DMs:
https://github.com/meshtastic/firmware/blob/70712d859cdc242b1873fa9b9622cf1c1711fb06/src/mesh/Router.cpp#L175-L179